File: Minimum_weights.py
import math
class Minimum_weights:
	def __init__(self,n):
		self.n = n
		self.count = 0
		self.weights = 0
	def print_weights(self):
		print("Minimum weights = ",self.count+self.weights)
		for i in range(self.count):
			print(3**i,end=" ")
		if self.weights!=0:
			print(self.weights,end=" ")
	def solve(self):
		max_3_pow = math.floor(math.log10(self.n)/math.log10(3))
		sum_3_series = int((3**(max_3_pow+1)-1)/2)
		self.count = max_3_pow+1
		self.weights=0 if (self.n-sum_3_series)<=0 else (self.n-sum_3_series)
		self.print_weights()
	def solve_unop(self):
		res=0
		count=0
		while True:
			pow_3 = 3**count
			res+=pow_3
			if pow_3>n:
				res-=pow_3
				break
			count+=1
		if self.n-res<=0:
			self.count=count
			return
		self.weights = self.n-res 
		self.count=count


if __name__=="__main__":
	n = int(input())
	Minimum = Minimum_weights(n)
	

File: Minimum_weights.c
#include<stdio.h>
#include<math.h>
#define MAX_WEIGHTS 1000
int input(int*);
int solve(int,int solution[]);
int output(int,int solution[]);
int main(){
	int n,count;
	int solution[MAX_WEIGHTS];
	input(&n);
	count = solve(n,solution);
	output(count,solution);
	return 0;
}
int input(int* n){
	scanf("%d",n);
}
int solve(int n,int solution[]){
	int res=0,count=0,i,pow_3;
	while(1){
		pow_3 = pow(3,count);
		res+=pow_3;
		if(pow_3>n){
			res-=pow_3;
			break;
		}
		solution[count]=pow_3;
		count++;
	}
	if((n-res)<=0){
		return count;
	}else{
		solution[count]=(n-res);
	}
	return count+1;
}
int output(int count,int solution[]){
	int i;
	printf("The minimum number of weights = %d\n",count);
	printf("The weights are :\n");
	printf("[ ");
	for(i=0;i<count;i++){
		printf("%d ",solution[i]);
	}
	printf("]");
}